docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+slightly fewer than the 175 in the hub. blst, a popular implementation of BLS
+signature aggregation was benchmarked to perform verification on 100 signatures in
+1.5 milliseconds [when run locally][blst-verify-bench] on an 8 thread machine and
+pre-aggregated public keys.
+
+#### Reduce Light-Client Verification Time
+
+The light client aims to be a faster and lighter-weight way to verify that a 
+block was voted on by a Tendermint network. The light client fetches
+Tendermint block headers and commit signatures, performing public key
+verification to ensure that the associated validator set signed the block.
+Reducing the size of the commit signature would allow the light client to fetch
+block data more quickly.
+
+Additionally, the faster signature verification times of BLS signatures mean
+that light client verification would proceed more quickly.
+
+#### Reduce Gossip Bandwidth
+
+##### Vote Gossip
+
+It is possible to aggregate signatures during voting and not need to gossip all 
+*n* validator signatures to all other validators. Theoretically, subsets of
+the signatures could be aggregated during consensus to produce vote messages
+carrying aggregated signatures. Implementations of this would certainly
+increase the complexity of the gossip layer but may possibly reduce the total
+number of signatures required to be verified by each validator.
+
+##### Block Gossip
+
+A reduction in the block size as a result of signature aggregation would
+naturally lead to a reduction in the bandwidth required to gossip a block.
+Each validator would only send and receive the smaller aggregated signatures
+instead of the full list of multi-signatures as we have them now.
+
+### What are the drawbacks to aggregated signatures?
+
+#### Heterogeneous key types cannot be aggregated
+
+Only one type of signature can be aggregated and our legacy signing schemes
+cannot be aggregated. In practice, this means that aggregated signatures can
+be created over the set of all validators that use BLS signatures and validators
+with alternative key types, such as Ed25519 must be included separately in
+blocks and votes.
+
+#### Many HSMs do not support aggregated signatures
+
+HSMs or 'Hardware Signing Modules' are a popular way to manage private keys.
+They provide additional security for key management and should be used when
+possible for storing highly sensitive private key material.
+
+Below is a list of popular HSMs along with their support for BLS signatures.
+
+* YubiKey
+  * [No support][yubi-key-bls-support]
+* Amazon Cloud HSM
+  * [No support][cloud-hsm-support]
+* Ledger
+  * [Lists support for the BLS12-381 curve][ledger-bls-announce],
+
+I cannot find support listed for Google Cloud, although perhaps it exists.
+
+## Feasibility of implementation
+
+### Can aggregated signatures be added as soft-upgrades?
+
+In my estimation, yes. With the implementation of proposer-based timestamps, 
```

Comment:
Your reasoning here makes sense, but what's the upgrade path for a validator who wants to swap signature schemes?

Reply:
Was considering describing this in the doc, but thought it may be a bit overkill.

Briefly, the validator set is modified as a result of the `FinalizeBlock` response. Specifically, the response contains a list of [ValidatorUpdates](https://github.com/tendermint/tendermint/blob/7243ef71e23f771624a6a6b3b8f55508895ae926/abci/types/types.pb.go#L3806) that replace the current validator set with the updated set. The way to update to the new key type would be to return a new validator in the `ValidatorUpdates` with a BLS key in place of the old validator with the old key type. I'm not certain how key rotation currently works in the cosmos SDK, but what ever scheme we chose would likely look pretty similar. 

Validators would rotate out their old key types for new BLS keys once the functionality was in place in Tendermint. As validators update to BLS keys, Tendermint would include more and more of the validator signatures in the new `AggregatedSignature` portion of the block.

Reply:
Ah, I see. Does that then imply that either (a) the node has to halt and restart after the last block before that change (so that it can come back up using the correct key and signing algorithm) or (b) that we have to have a rule somewhere to say which signature scheme is being used at which height?

Or if neither of these is right, how does consensus know which key to use?

Reply:
I was considering that we could allow both simultaneously and allow transfer between one to the other. The node would maintain BLS public keys along with Ed25519 and Secp256k1 keys. It would continue to verify old signatures in the old way, one by one, and, separately, it would verify the whole BLS signature at once.

Reply:
Signing scheme could check for presence in both:
```
message Commit{
...
  repeated CommitSig signature = 4;
  AggregatedSignature aggregate_signature = 5;
...
}
```

Reply:
> I was considering that we could allow both simultaneously and allow transfer between one to the other. The node would maintain BLS public keys along with Ed25519 and Secp256k1 keys. It would continue to verify old signatures in the old way, one by one, and, separately, it would verify the whole BLS signature at once.

Right—clearly we still need the ability to verify old signatures regardless, though I guess we shouldn't need to generate new ones and could throw away the private key. So is the idea that if I _have_ a BLS key (at all) I use that for signing any new messages, from that moment forward?

I'm mainly wondering about the transition point: If the operator installs a BLS key "early" (before the network swaps), how do I know when it's time to start using it?

Reply:
Huzzah! I see what you're asking. How does the local node determine what key to sign with and/or how can it gracefully swap over to that new key. I feel like, to the extent to which that problem will exist in BLS signing, it will already exist now? I can go take a look at the code to see how updating to a new key may be done.

Reply:
As of now, updating a key would require switching out the new key and restarting Tendermint. There is no clear update path.